tools/lsp-wake/astree.h

```diff
@@ -63,24 +64,46 @@ class ASTree {
   struct SymbolUsage {
     Location usage;
     Location definition;
-    SymbolUsage(Location _usage, Location _definition): usage(std::move(_usage)), definition(std::move(_definition)) {}
-
+    SymbolUsage(Location _usage, Location _definition)
+        : usage(std::move(_usage)), definition(std::move(_definition)) {}
   };
 
   struct Comment {
-    Comment(): location("") {}
-    Comment(std::string _comment_text, Location _location, int _level): comment_text(std::move(_comment_text)), location(std::move(_location)), level(_level) {}
+    Comment() : location("") {}
+    Comment(std::string _comment_text, Location _location, int _level)
+        : comment_text(std::move(_comment_text)), location(std::move(_location)), level(_level) {}
 
     std::string comment_text;
     Location location;
     int level;  // level of nestedness in the tree
   };
 
+  struct LineLocation {
+    std::string filename;
+    int row;
+
+    explicit LineLocation(const Location &location)
+        : filename(location.filename), row(location.start.row) {}
+    LineLocation(std::string _filename, int _row) : filename(std::move(_filename)), row(_row) {}
+  };
+
+  struct HashLineLocation {
+    std::size_t operator()(const LineLocation &loc) const {
+      return wcl::hash_combine(std::hash<std::string>{}(loc.filename), std::hash<int>{}(loc.row));
+    }
+  };
+
+  struct EqualLineLocation {
+    bool operator()(const LineLocation &lhs, const LineLocation &rhs) const {
+      return lhs.row == rhs.row && lhs.filename == rhs.filename;
+    }
+  };
```

Comment:
Not worth the change here but in the future you could have implemented the `==` operator inside the class to keep the default equality struct. Neither is better than the other really.